shared/json_utilities.go

```diff
@@ -0,0 +1,197 @@
+	// Check if all elements of arrOne exist in arrTwo
+	for _, elementOne := range arrOne {
+		found := false
+		for _, elementTwo := range arrTwo {
+			if IsSubset(elementOne, elementTwo) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+// Function to check if a string contains regex-specific characters
+func isPotentialRegex(s string) bool {
+	// Regex characters to check for
+	regexChars := []string{".", "*", "+", "?", "^", "$", "[", "]", "(", ")", "|", "{", "}"}
+
+	// Check if the string contains any of the characters used in regular expressions
+	for _, char := range regexChars {
+		if strings.Contains(s, char) {
+			return true
+		}
+	}
+	return false
+}
+
+// Helper function to compare strings. Since mock definitions support
+// regex, we need to check if the string is a regex and if so, compare it
+func StringCompare(patternOrStr, str string) bool {
+	// Try to compile the pattern to check if it's a valid regex
+	if isPotentialRegex(patternOrStr) {
+		if _, err := regexp.Compile(patternOrStr); err != nil {
+			// If it's not a valid regex, do a normal string comparison
+			return patternOrStr == str
+		}
+
+		// If it's a valid regex, check if it matches the string
+		matched, err := regexp.MatchString(patternOrStr, str)
+		if err != nil {
+			return false
+		}
+
+		return matched
+	}
+
+	// If it's not a regex, do a normal string comparison
+	return patternOrStr == str
+}
+
+// Helper function to get the value based on a JSON path (dot notation or array index).
+func getValueByPath(data interface{}, path string) (interface{}, error) {
+	// Split the path by dots and array indices (i.e., "[7]")
+	re := regexp.MustCompile(`\.`)
+	parts := re.Split(path, -1)
+
+	// Traverse the path
+	for _, part := range parts {
+		if part == "" {
+			continue
+		}
+		if strings.HasPrefix(part, "[") && strings.HasSuffix(part, "]") {
+			// Handle array access, e.g., "[7]"
+			// Get the array index
+			index := part[1 : len(part)-1]
+			// Traverse the array using the index
+			if idx, err := strconv.Atoi(index); err == nil {
+				if arr, ok := data.([]interface{}); ok && idx < len(arr) {
+					data = arr[idx]
+				} else {
+					return nil, fmt.Errorf("invalid array index or not an array")
+				}
+			} else {
+				return nil, fmt.Errorf("invalid array index: %s", part)
+			}
+		} else {
+			// Handle map key access
+			if m, ok := data.(map[string]interface{}); ok {
+				if val, found := m[part]; found {
+					data = val
+				} else {
+					return nil, fmt.Errorf("key not found: %s", part)
+				}
+			} else {
+				return nil, fmt.Errorf("not a map: %v", data)
+			}
+		}
+	}
+
+	return data, nil
+}
+
+// Function to replace template variables in JSON path format
+func ReplaceTemplateVars(jsonStr string, vars interface{}) (string, error) {
+	// Regular expression to match the ${var} format (full path)
+	re := regexp.MustCompile(`\$\{([a-zA-Z0-9._\[\]]+)\}`)
```

Comment:
it's always a good idea to pre-compile all regex OUTSIDE of the function it's called from. Why? it's REALLY SLOW to compile regex in realtime, define this as a module variable, compile it using `Init()` and then every time the `ReplaceTemplateVars` is called, it will run s fast as possible.